### PR TITLE
Control Statements

### PR DESCRIPTION
**Pull request by joerivanarkel**

### CODE REVIEW

**Comment by joerivanarkel** [comment]
**Summary**:
- Added `IfStatementNode` and Parser
- Additional `Token.Matches()` test after Value check fix

File: Common/FileHandlers/FileReadHandler.cs [hunk 1]
```diff
@@ -24,11 +25,11 @@ public async Task<string> ReadAllTextFromFile(string name, string extension, str
...
     }
 
-    public bool FileExists(string name, string extension, string location)
+    public bool FileExists(string name, FileExtension extension, string location)
     {
-        if (name == string.Empty || extension == string.Empty) { throw new ArgumentNullException("Name, Text or Extension provided is empty"); }
+        if (name == string.Empty) { throw new ArgumentNullException("Name, Text or Extension provided is empty"); }
 
-        name += extension;
+        name += "." + extension.ToString().ToLower();
```
Comment (teusvanarkel) [inline Common/FileHandlers/FileReadHandler.cs:32]: Make a createFileName method

Comment (teusvanarkel) [reply Common/FileHandlers/FileReadHandler.cs:32]: Make the method in a base FileHandler class so you can use it in both FileReadHandler and FileWriteHandler

Comment (teusvanarkel) [reply Common/FileHandlers/FileReadHandler.cs:32]: Make a UnitTest for it

File: Common/Logger.cs [hunk 2]
```diff
@@ -23,16 +23,16 @@ public Logger(IFileWriteHandler FileWriteHandler, IEnvironmentVariables environm
             logFileName = $"{DateTime.Now.Day}.{DateTime.Now.Month}.{DateTime.Now.Year}.{DateTime.Now.Hour}.{DateTime.Now.Minute}.{DateTime.Now.Second}";
         }
 
-        location = $"bin/pirate{version}/logs";
         version = environmentVariables.GetVariable("version");
```
Comment (teusvanarkel) [inline Common/Logger.cs:26]: What happens if there is no version in the config....maybe make a default version?

File: Common/Logger.cs [hunk 3]
```diff
@@ -23,16 +23,16 @@ public Logger(IFileWriteHandler FileWriteHandler, IEnvironmentVariables environm
...
+        location = $"bin/pirate{version}/logs";
     }
 
     public bool Log(string message, string orginFile, LogType logType)
     {
         var time = DateTime.Now.ToString();
         var text = ($"{time.Replace(" uur", "")}: {logType.ToString()}: {orginFile}.cs: {message}\n");
 
-        _fileWriteHandler.AppendToFile(new FileWriteModel("log", ".log", location, text));
+        _fileWriteHandler.AppendToFile(new FileWriteModel(logFileName, FileExtension.LOG, location, text));
```
Comment (teusvanarkel) [inline Common/Logger.cs:35]: Make fileExtension enum lower caps....then in the filehandlers it doesnt need to be lowercased

File: PirateInterpreter/Interpreter.cs [hunk 4]
```diff
@@ -29,7 +29,10 @@ public List<BaseValue> StartInterpreter(string filename)
         {
             Logger.Log($"Interpreting {item.GetType().Name}", this.GetType().Name, LogType.INFO);
             var interpreter = interpreterFactory.GetInterpreter(item, Logger);
-            result.Add(interpreter.VisitNode());
+            foreach (var value in interpreter.VisitNode())
+            {
+                result.Add(value);
```
Comment (teusvanarkel) [inline PirateInterpreter/Interpreter.cs:34]: Maybe addRange is possible ( then the foreach can be removed )

File: PirateInterpreter/Interpreters/ComparisonOperationNodeInterpreter.cs [hunk 5]
```diff
@@ -13,66 +13,70 @@ public ComparisonOperationNodeInterpreter(INode node, InterpreterFactory Interpr
...
         var interpreter = InterpreterFactory.GetInterpreter(Node.Left, Logger );
-        var left = interpreter.VisitNode();
+        var leftNode = interpreter.VisitNode();
+
+        if (leftNode.Count > 1) throw new Exception("Left value is not a single value");
+        var left = leftNode[0];
 
         interpreter = InterpreterFactory.GetInterpreter(Node.Right, Logger);
-        var Right = interpreter.VisitNode();  
+        var rightNode = interpreter.VisitNode();
```
Comment (teusvanarkel) [inline PirateInterpreter/Interpreters/ComparisonOperationNodeInterpreter.cs:25]: Make a base class method and put all the clutter there.

File: PirateInterpreter/Interpreters/ForLoopStatementNodeInterpreter.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using PirateInterpreter.Values;
+
+namespace PirateInterpreter.Interpreters;
+
+public class ForLoopStatementNodeInterpreter : BaseInterpreter
+{
+    public IForLoopStatementNode Node { get; set; }
```
Comment (teusvanarkel) [inline PirateInterpreter/Interpreters/ForLoopStatementNodeInterpreter.cs:11]: Is this used...why not naming: ForLoopStatementNode

File: PirateInterpreter/Interpreters/ForLoopStatementNodeInterpreter.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+    public ForLoopStatementNodeInterpreter(INode node, InterpreterFactory InterpreterFactory, ILogger logger) : base(logger, InterpreterFactory)
```
Comment (teusvanarkel) [inline PirateInterpreter/Interpreters/ForLoopStatementNodeInterpreter.cs:13]: Ask in constructor for interpreterFactory with its interface

File: PirateInterpreter/Interpreters/ForLoopStatementNodeInterpreter.cs [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        if (node is not IForLoopStatementNode) throw new TypeConversionException(node.GetType(), typeof(IIfStatementNode));
+        Node = (IForLoopStatementNode)node;
+
+        Logger.Log($"Created {this.GetType().Name} : \"{Node.ToString()}\"", this.GetType().Name, Common.Enum.LogType.INFO);
+    }
+    public override List<BaseValue> VisitNode()
+    {
+        var variableValueNode = InterpreterFactory.GetInterpreter(Node.VariableNode, Logger).VisitNode();
```
Comment (teusvanarkel) [inline PirateInterpreter/Interpreters/ForLoopStatementNodeInterpreter.cs:22]: One call per line for readability

File: PirateInterpreter/Interpreters/BaseInterpreter.cs [hunk 9]
```diff
@@ -4,7 +4,6 @@ namespace PirateInterpreter.Interpreters;
 
 public abstract class BaseInterpreter
 {
-    public abstract BaseValue VisitNode();
     protected ILogger Logger {get; set; }
```
Comment (teusvanarkel) [inline PirateInterpreter/Interpreters/BaseInterpreter.cs:7]: Make Logger a readOnly protected field

File: PirateInterpreter/Values/Variable.cs [hunk 10]
```diff
@@ -17,9 +17,17 @@ public object GetValue(string value)
         var resultNode = SymbolTable.Instance(Logger).Get(value);
         if (resultNode is not IValueNode)
         {
-           var interpreter = InterpreterFactory.GetInterpreter(resultNode, Logger);
-           var result = interpreter.VisitNode();
-           return result.Value;
+
+            var interpreter = InterpreterFactory.GetInterpreter(resultNode, Logger);
+            var result = interpreter.VisitNode();
+            if (result.Count == 1)
```
Comment (teusvanarkel) [inline PirateInterpreter/Values/Variable.cs:23]: Check out the ternary operator

File: PirateLexer/Tokens/Token.cs [hunk 11]
```diff
@@ -19,7 +19,7 @@ public Token(TokenGroup tokenGroup, object tokenType, object value = null)
 
     public bool Matches(object tokenType, object value=null)
     {
-        if (Value == null)
+        if (value == null || Value == null)
         {
             return TokenType.Equals(tokenType);
         }
```
Comment (teusvanarkel) [inline PirateLexer/Tokens/Token.cs:25]: Dont use DUTCH in you code

File: PirateParser/Node/ForLoopStatementNode.cs [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+    public override string ToString()
+    {
+        string resultString = string.Empty;
+
+        foreach (var node in BodyNodes)
+        {
+            resultString += node.ToString() + '\n';
+        }
+        return $"if ({VariableNode.ToString()}) \n{{ \n {string.Join(" ", BodyNodes)} \n}}";
```
Comment (teusvanarkel) [inline PirateParser/Node/ForLoopStatementNode.cs:44]: Change the return string ( if must be for )

File: PirateParser/Node/IfStatementNode.cs [hunk 13]
```diff
@@ -0,0 +1,48 @@
+using PirateParser.Node.Interfaces;
+
+namespace PirateParser.Node;
+
+[Serializable]
+public class IfStatementNode : IIfStatementNode
+{
+    public IOperationNode ConditionNode { get; set; }
+    public List<INode> BodyNodes { get; set; }
+    public List<INode> ElseNode { get; set; }
```
Comment (teusvanarkel) [inline PirateParser/Node/IfStatementNode.cs:10]: MAKE EVERYTHING .NET 7

File: PirateParser/Parsers/ForLoopStatementParser.cs [hunk 14]
```diff
@@ -0,0 +1,78 @@
+using PirateParser.Node;
+using PirateParser.Node.Interfaces;
+
+namespace PirateParser.Parsers;
+
+public class ForLoopStatementParser : BaseParser
+{
+    private ParserFactory _parserFactory { get; set; }
+
+    public ForLoopStatementParser(List<Token> tokens, Token token, ILogger logger, ParserFactory parserFactory) : base(tokens, token, logger)
```
Comment (teusvanarkel) [inline PirateParser/Parsers/ForLoopStatementParser.cs:10]: Think if its possible to pass in the list and define the current one at the same time.

**Pull request merged**